src/mlpack/tests/main_tests/kfn_test.cpp

```diff
@@ -0,0 +1,378 @@
+  SetInputParam("query", queryData);
+  SetInputParam("k", (int) 10);
+
+  mlpackMain();
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  neighbors = std::move(CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  distances = std::move(CLI::GetParam<arma::mat>("distances"));
+  
+  // Reset passed parameters. 
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["query"].wasPassed = false;
+
+  // Input saved model, pass the same query and keep k unchanged.
+  SetInputParam("input_model", 
+      std::move(CLI::GetParam<KFNModel*>("output_model")));
+  SetInputParam("query", queryData);
+  
+  mlpackMain();
+
+  // Check that initial output matrices and the output matrices using 
+  // saved model are equal.
+  CheckMatrices(neighbors, CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  CheckMatrices(distances, CLI::GetParam<arma::mat>("distances"));
+}
+
+/*
+ * Ensure that different search algorithms give same result.
+ */
+BOOST_AUTO_TEST_CASE(KFNAllAlgorithmsTest)
+{ 
+  string algorithms[] = {"dual_tree", "naive", "single_tree", "greedy"};
+  int nof_algorithms = sizeof(algorithms)/sizeof(algorithms[0]);
+
+  // Neighbors and distances given by the above algorithms will be stored 
+  // in the following arrays in the order:
+  // dual_tree, naive, single_tree, greedy. 
+  arma::Mat<size_t> neighbors[nof_algorithms]; 
+  arma::mat distances[nof_algorithms];
+
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  arma::mat queryData;
+  queryData.randu(3, 90); // 90 points in 3 dimensions.
+
+  // Keep some k <= number of reference points same over all.
+  SetInputParam("k", (int) 10);
+
+  // Looping over all the algorithms and storing their outputs.
+  for (int i = 0; i < nof_algorithms; i++) 
+  {
+    // Same random inputs, different algorithms.
+    SetInputParam("reference", referenceData);
+    SetInputParam("query", queryData);
+    SetInputParam("algorithm", algorithms[i]);
+
+    mlpackMain();
+
+    neighbors[i] = std::move(CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+    distances[i] = std::move(CLI::GetParam<arma::mat>("distances"));
+
+    CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+    CLI::GetSingleton().Parameters()["query"].wasPassed = false;
+    CLI::GetSingleton().Parameters()["algorithm"].wasPassed = false;
+  }
+  
+  // The looped matrices check fails.
+  // The following individual check passes with greedy outputs
+  // out.
+  // for (int i = 0; i < nof_algorithms - 1; i++)
+  // {
+  //   CheckMatrices(neighbors[i], neighbors[i + 1]);
+  //   CheckMatrices(distances[i], distances[i + 1]);
+  // }
+  CheckMatrices(neighbors[0], neighbors[1]);
+  CheckMatrices(neighbors[1], neighbors[2]);
+  //CheckMatrices(neighbors[2], neighbors[3]);
+  CheckMatrices(distances[0], distances[1]);
+  CheckMatrices(distances[1], distances[2]);
+  //CheckMatrices(distances[2], distances[3]);
```

Comment:
Yep, it's fine to avoid checking the greedy output here.  I think it might be useful just to make sure there are no `DBL_MAX`es in the `distances[3]` matrix though.